File: front sources/com/qx/front/io/webgl/shapes/WebGL_Appearance.js

/**
 * CAD_Engine
 * 
 */
function WebGL_Appearance(){

	// Material
	
	/** the default program for wire rendering */
	this.wireProgramId = "color2";
	
	/** default value for shape material */
	this.wireColor = [0.12, 0.12, 0.12, 1.0];

	/** the default program for wire rendering */
	this.surfaceProgramId = "standard";
	
	/** default value for shape material -> "Standard" Unity-style shading */
	this.surfaceGlossiness = 0.7;
	
	/** default value for shape material -> "Standard" Unity-style shading */
	this.surfaceRoughness = 0.5;
	
	/** default value for shape material -> Phong shading */
	this.surfaceShininess = 0.5;
	
	/** default value for shape material -> multi-purposes */
	this.surfaceSpecularColor = [0.12, 0.8, 0.8, 1.0];

	/** default value for shape material -> multi-purposes */
	this.surfaceDiffuseColor = [0.12, 0.8, 0.8, 1.0];

	/** default value for shape material -> multi-purposes */
	this.surfaceAmbientColor = [0.1, 0.1, 0.1, 1.0];
	
	/** the textures */
	this.surfaceTexture0 = null;
	this.surfaceTexture1 = null;
	this.surfaceTexture2 = null;
	this.surfaceTexture3 = null;
	
}


WebGL_Appearance.prototype = {

		apply : function(instance){
			
			// material part
			if(instance.isWireEnabled){
				instance.wireColor = this.wireColor;		
				
				// finally set program
				instance.setWireProgram(this.wireProgramId);
			}

			// surface
			if(instance.isSurfaceEnabled){
				
				instance.surfaceGlossiness = this.surfaceGlossiness;
				instance.surfaceRoughness = this.surfaceRoughness;
				instance.surfaceShininess = this.surfaceShininess;
				instance.surfaceSpecularColor = this.surfaceSpecularColor;
				instance.surfaceDiffuseColor = this.surfaceDiffuseColor;
				instance.surfaceAmbientColor = this.surfaceAmbientColor;

				// textures
				instance.surfaceTexture0 = this.surfaceTexture0;
				instance.surfaceTexture1 = this.surfaceTexture1;
				instance.surfaceTexture2 = this.surfaceTexture2;
				instance.surfaceTexture3 = this.surfaceTexture3;	
				
				// prgm
				instance.setSurfaceProgram(this.surfaceProgramId);
			}
		}
};

File: back sources/com/qx/back/io/webgl/WebGL_Appearance.java
package com.qx.back.io.webgl;

import java.io.IOException;

import com.qx.back.base.bytes.ByteOutput;
import com.qx.back.io.bohr.BohrObject;
import com.qx.back.io.bohr.BohrScope;

public class WebGL_Appearance extends BohrObject {

	/**
	 * default program for wire rendering 
	 */
	public String wireProgram = "color2";

	/**
	 * default value for shape material
	 */
	public WebGL_RGBA8Color wireColor = new WebGL_RGBA8Color(76, 76, 76, 255);

	/**
	 * default program for surface rendering 
	 */
	public String surfaceProgram = "standard";

	/**
	 * default value for shape material
	 */
	public double surfaceGlossiness = 0.7;


	/**
	 * default value for shape material
	 */
	public double surfaceRoughness = 0.5;


	/**
	 * Phong exponent for Old-style rendering
	 */
	public double surfaceShininess = 2.5;


	/** 
	 * default value for shape material
	 */
	public WebGL_RGBA8Color surfaceSpecularColor = new WebGL_RGBA8Color(255, 255, 255, 255);


	/** 
	 * default value for shape material
	 */
	public WebGL_RGBA8Color surfaceDiffuseColor = new WebGL_RGBA8Color(255, 255, 255, 255);


	/** 
	 * default value for shape material
	 */
	public WebGL_RGBA8Color surfaceAmbientColor = new WebGL_RGBA8Color(255, 255, 255, 255);


	/**
	 * surface color texture
	 */
	public String surfaceTexture0 = null;

	/**
	 * surface color texture
	 */
	public String surfaceTexture1 = null;

	/**
	 * surface color texture
	 */
	public String surfaceTexture2 = null;

	/**
	 * surface color texture
	 */
	public String surfaceTexture3 = null;




	public final static int TYPE_CODE = 201;

	@Override
	public int getTypeCode() {
		return WebGL_Back.WEBGL_BOHR_PREFIX+TYPE_CODE;
	}

	@Override
	public void send(ByteOutput output) throws IOException {

		// material

		// <wire>
		output.putStringUTF8(wireProgram);
		wireColor.write(output);
		// </wire>

		// <surface>
		output.putStringUTF8(surfaceProgram);
		output.putFloat32((float) surfaceGlossiness);
		output.putFloat32((float) surfaceRoughness);
		output.putFloat32((float) surfaceShininess);
		surfaceSpecularColor.write(output);
		surfaceDiffuseColor.write(output);
		surfaceAmbientColor.write(output);

		// textures
		output.putStringUTF8(surfaceTexture0);
		output.putStringUTF8(surfaceTexture1);
		output.putStringUTF8(surfaceTexture2);
		output.putStringUTF8(surfaceTexture3);
		// </surface>
	}

	@Override
	public void getSubNodes(BohrScope scope) {
		// none
	}

}
